examples/merge.c

```diff
@@ -0,0 +1,387 @@
+
+	checkout_opts.checkout_strategy = GIT_CHECKOUT_FORCE|GIT_CHECKOUT_ALLOW_CONFLICTS;
+
+	err = git_merge_analysis(&analysis, &preference,
+	                         repo,
+	                         (const git_annotated_commit **)opts->annotated,
+	                         opts->annotated_count);
+	check_lg2(err, "merge analysis failed", NULL);
+
+	if (analysis & GIT_MERGE_ANALYSIS_UP_TO_DATE) {
+		printf("Already up-to-date\n");
+		return 0;
+	} else if (analysis & GIT_MERGE_ANALYSIS_UNBORN ||
+		       (analysis & GIT_MERGE_ANALYSIS_FASTFORWARD &&
+		       !(preference & GIT_MERGE_PREFERENCE_NO_FASTFORWARD))) {
+		const git_oid *target_oid;
+		if (analysis & GIT_MERGE_ANALYSIS_UNBORN) {
+			printf("Unborn\n");
+		} else {
+			printf("Fast-forward\n");
+		}
+
+		/* Since this is a fast-forward, there can be only one merge head */
+		target_oid = git_annotated_commit_id(opts->annotated[0]);
+
+		return perform_fastforward(repo, target_oid, (analysis & GIT_MERGE_ANALYSIS_UNBORN));
+	} else if (analysis & GIT_MERGE_ANALYSIS_NORMAL) {
+		if (preference & GIT_MERGE_PREFERENCE_FASTFORWARD_ONLY) {
+			printf("Fast-forward is preferred, but only a merge is possible\n");
+			return -1;
+		}
+
+		err = git_merge(repo,
+		                (const git_annotated_commit **)opts->annotated, opts->annotated_count,
+		                &merge_opts, &checkout_opts);
+		if (err != 0)
+			return -1;
+
+		/* Inform that a merge was done */
+		opts->did_merge = 1;
+
+		return 0;
+	}
+
+	return -1;
+}
+
+int main(int argc, char **argv)
+{
+	git_repository *repo = NULL;
+	merge_options opts;
+	git_index *index;
+	git_repository_state_t state;
+	const char *path = ".";
+	int err = 0;
+
+	merge_options_init(&opts);
+	parse_options(&path, &opts, argc, argv);
+
+	git_libgit2_init();
+
+	check_lg2(git_repository_open_ext(&repo, path, 0, NULL),
+	          "Could not open repository", NULL);
+
+	state = git_repository_state(repo);
+	if (state != GIT_REPOSITORY_STATE_NONE) {
+		fprintf(stderr, "repository is in unexpected state %d\n", state);
+		goto cleanup;
+	}
+
+	err = resolve_heads(repo, &opts);
+	if (err != 0)
+		goto cleanup;
+
+	err = analyze_merge(repo, &opts);
+	if (err != 0) {
+		fprintf(stderr, "merge failed\n");
+		goto cleanup;
+	}
+
+	if (!opts.did_merge) {
+		/* Was either up-to-date, unborn, or a fast-forward, nothing left to do */
+		goto cleanup;
+	}
+
+	check_lg2(git_repository_index(&index, repo), "failed to get repository index", NULL);
+
+	if (git_index_has_conflicts(index)) {
+		/* Handle conflicts */
+		git_index_conflict_iterator *conflicts;
+		const git_index_entry *ancestor;
+		const git_index_entry *our;
+		const git_index_entry *their;
+
+		check_lg2(git_index_conflict_iterator_new(&conflicts, index), "failed to create conflict iterator", NULL);
+
+		while ((err = git_index_conflict_next(&ancestor, &our, &their, conflicts)) == 0) {
+			fprintf(stderr, "conflict: a:%s o:%s t:%s\n",
+			        ancestor ? ancestor->path : "", our->path, their->path);
```

Comment:
`our` and `their` can also be `NULL` in case of add-remove, modify-remove or similar merge conflicts.